br/pkg/lightning/backend/local/duplicate.go

```diff
@@ -180,645 +234,570 @@ func physicalTableIDs(tableInfo *model.TableInfo) []int64 {
-	decoder *kv.TableKVDecoder,
-	req *DuplicateRequest,
-	hasDupe *atomic.Bool,
-) error {
-	logger := log.With(
-		zap.String("table", decoder.Name()),
-		zap.Int64("tableID", req.tableID),
-		logutil.Key("startKey", req.start),
-		logutil.Key("endKey", req.end))
+type DupKVStream interface {
+	Next() (key, val []byte, err error)
+	Close() error
+}
 
-	startKey := codec.EncodeBytes([]byte{}, req.start)
-	endKey := codec.EncodeBytes([]byte{}, req.end)
+//goland:noinspection GoNameStartsWithPackageName
+type LocalDupKVStream struct {
+	iter pkgkv.Iter
+}
 
-	regions, err := restore.PaginateScanRegion(ctx, manager.splitCli, startKey, endKey, scanRegionLimit)
-	if err != nil {
-		return err
+func NewLocalDupKVStream(dupDB *pebble.DB, keyAdapter KeyAdapter, keyRange tidbkv.KeyRange) *LocalDupKVStream {
+	opts := &pebble.IterOptions{
+		LowerBound: keyRange.StartKey,
+		UpperBound: keyRange.EndKey,
 	}
-	tryTimes := 0
-	indexHandles := makePendingIndexHandlesWithCapacity(0)
-	for len(regions) > 0 {
-		if tryTimes > maxRetryTimes {
-			return errors.Errorf("retry time exceed limit")
-		}
-		unfinishedRegions := make([]*restore.RegionInfo, 0)
-		waitingClients := make([]import_sstpb.ImportSST_DuplicateDetectClient, 0)
-		watingRegions := make([]*restore.RegionInfo, 0)
-		for idx, region := range regions {
-			if len(waitingClients) > manager.regionConcurrency {
-				r := regions[idx:]
-				unfinishedRegions = append(unfinishedRegions, r...)
-				break
-			}
-			_, start, _ := codec.DecodeBytes(region.Region.StartKey, []byte{})
-			_, end, _ := codec.DecodeBytes(region.Region.EndKey, []byte{})
-			if bytes.Compare(startKey, region.Region.StartKey) > 0 {
-				start = req.start
-			}
-			if region.Region.EndKey == nil || len(region.Region.EndKey) == 0 || bytes.Compare(endKey, region.Region.EndKey) < 0 {
-				end = req.end
-			}
-
-			logger.Debug("[detect-dupe] get duplicate stream",
-				zap.Int("localStreamID", idx),
-				logutil.Region(region.Region),
-				logutil.Leader(region.Leader),
-				logutil.Key("regionStartKey", start),
-				logutil.Key("regionEndKey", end))
-			cli, err := manager.getDuplicateStream(ctx, region, start, end)
-			if err != nil {
-				r, err := manager.splitCli.GetRegionByID(ctx, region.Region.GetId())
-				if err != nil {
-					unfinishedRegions = append(unfinishedRegions, region)
-				} else {
-					unfinishedRegions = append(unfinishedRegions, r)
-				}
-			} else {
-				waitingClients = append(waitingClients, cli)
-				watingRegions = append(watingRegions, region)
-			}
-		}
+	iter := newDupDBIter(dupDB, keyAdapter, opts)
+	iter.First()
+	return &LocalDupKVStream{iter: iter}
+}
 
-		if indexHandles.Len() > 0 {
-			handles := manager.getValues(ctx, decoder, indexHandles)
-			if handles.Len() > 0 {
-				indexHandles = handles
-			} else {
-				indexHandles.truncate()
-			}
+func (s *LocalDupKVStream) Next() (key, val []byte, err error) {
+	if !s.iter.Valid() {
+		err = s.iter.Error()
+		if err == nil {
+			err = io.EOF
```

Comment:
I prefer to check error for the end of iteration explicitly, rather than checking the return slice. This prevents the caller from using the slice incorrectly. See also https://github.com/golang/go/wiki/CodeReviewComments#in-band-errors.